### ISSUE TITLE
CredSSP support

### ISSUE THREAD

**Issue by semancik** [cloudsoft/winrm4j#49]
Hi,
I have just implemented support for CredSSP in Apache HTTP client:
https://github.com/Evolveum/httpclient/tree/credssp
and did some basic testing with winrm4j:
https://github.com/Evolveum/winrm4j/tree/credssp
These are my private forks/branches for now. I'm mentioning that for anybody interested in joining the testing. The plan is to contribute the CredSSP code to the Apache HTTP components project. Then it could be cleanly used in winrm4j.
Sorry for opening an issues here. But there seems to be no other way how to communicate with you guys (no project page, no mailing list, no mail addresses on github profile)

**Comment by neykov** [comment]
That's awesome work @semancik. Thanks for the heads up.

Opening it as an issue is actually useful - will let other users see that it will be supported at some point.
Good point about missing contact info - we will address this.

**Comment by bostko** [comment]
@semancik I think the best way to contribute and communicate your changes is via Pull Request.
I made one for you https://github.com/cloudsoft/winrm4j/pull/50

**Comment by semancik** [comment]
Thank you @bostko. But I know how pull requests work, of course. The problem is that this pull request does not make any sense now. If it will get merged then winrm4j won't compile. Because the HTTPC contribution is not merged yet. And even if it was merged, we need to wait for HTTPC release to update the winrm4j dependency. And even that is not that easy because this is transitive dependency through CXF. This process can take months to complete. Therefore I would not like to issue a pull request to Apache unless this code is tested at least a bit more. That was the primary purpose of this communication: to attract more people from the winrm4j community to test the code.

**Comment by semancik** [comment]
Thank you @bostko. But I know how pull requests work, of course. The problem is that this pull request does not make any sense now. If it will get merged then winrm4j won't compile. Because the HTTPC contribution is not merged yet. And even if it was merged, we need to wait for HTTPC release to update the winrm4j dependency. And even that is not that easy because this is transitive dependency through CXF. This process can take months to complete. Therefore I would not like to issue a pull request to Apache unless this code is tested at least a bit more. That was the primary purpose of this communication: to attract more people from the winrm4j community to test the code.

**Comment by semancik** [comment]
Update: My CredSSP contribution to the HTTP Client will probably not be accepted:
https://github.com/apache/httpclient/pull/66
I'm sorry guys. I have spend more than 100 working hours on this implementation. I have dealt with obsolete crypto, terrible Microsoft documentation, absolutely no means to diagnose the issues on the Windows side. But it works. The implementation is done and tested. But my patience with the contribution dance has run out. If anyone of you have the time to deal with this then please take over and finish that bureaucratic part. I won't. I can't do that any more. I'm sorry.

**Comment by neykov** [comment]
That's a bummer @semancik. I know the feeling - there's unexpectedly large effort to get code merged in OSS projects.

Looks like formatting would be straightforward to fix here, definitely something I can help with. The other big issue seems to be missing tests. I'm afraid I'm lost as to how the internals of the CredSSP protocol works. Do you think you could do something about that? Would be a waste not to have this merged.

**Comment by semancik** [comment]
Yeah. I can help with the tests. However, it would be very hard to create end-to-end tests for CredSSP. That would require interactive server implemenation - in fact a TLS server. It is possible, but it may take several days. I simply do not have that time. But adding few tests for the NTLM signing should be possible. I can do these quite easily. And as far as I understood that's what the Apache people are looking for.
The problem for me is the communication. They ask to reformat code without specifying what is the correct formating on the web site. I have already reformated the code. But it turned out to be the wrong style. Then they ask for the tests. If we supply the tests what will they ask then? Is it worth the effort? But if you are willing to put it through I can promise to write the NTLM tests. However, I cannot promise CredSSP tests.

Just a side note: I'm maintaining a fairly large open source project myself. But I would never dream of putting obstacles like this for someone who is willing to contribute interesting piece of code. I would do anything to make the contribution possible. I just cannot understand the behavior of these Apache guys. And I'm really surprised and upset about this ... anyway ... if you are willing to help with this then it is really appreciated.

**Comment by neykov** [comment]
Update:
The `httpcomponents-client` changes were merged in March. This is blocked on getting a new release of the library out. Ideally we'll be able to use it right after release, overriding the version coming from CXF. If there are any incompatibilities though will need to wait for CXF to bump it's dependency version and do a release of its own.
